plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenPath.java

```diff
@@ -0,0 +1,377 @@
+    }
+
+    @Nonnull
+    public String getExtension() {
+      return extension;
+    }
+
+    @Nullable
+    public SignatureType getSignatureType() {
+      return signatureType;
+    }
+  }
+
+  private final String path;
+
+  private final String fileName;
+
+  private final HashType hashType;
+
+  private final Coordinates coordinates;
+
+  public MavenPath(final String path, final Coordinates coordinates)
+  {
+    this.path = checkNotNull(path);
+    this.fileName = this.path.substring(path.lastIndexOf('/') + 1);
+    HashType ht = null;
+    for (HashType v : HashType.values()) {
+      if (this.fileName.endsWith("." + v.getExt())) {
+        ht = v;
+        break;
+      }
+    }
+    this.hashType = ht;
+    this.coordinates = coordinates;
+  }
+
+  @Nonnull
+  public Kind getKind() {
+    if (getCoordinates() != null) {
+      return Kind.ARTIFACT;
+    }
+    else if (main().getFileName().equals(MAVEN_METADATA_FILENAME)) {
+      return Kind.METADATA;
+    }
+    else {
+      return Kind.OTHER;
+    }
+  }
+
+  @Nonnull
+  public String getPath() {
+    return path;
+  }
+
+  @Nonnull
+  public String getFileName() {
+    return fileName;
+  }
+
+  @Nullable
+  public HashType getHashType() {
+    return hashType;
+  }
+
+  @Nullable
+  public Coordinates getCoordinates() { return coordinates; }
+
+  /**
+   * Returns {@code true} if this path is subordinate of another path.
+   *
+   * @see {@link #subordinateOf()}
+   */
+  public boolean isSubordinate() {
+    return isHash() || isSignature();
+  }
+
+  /**
+   * Returns {@code true} if this path represents a hash.
+   */
+  public boolean isHash() {
+    return hashType != null;
+  }
+
+  /**
+   * Returns {@code true} if this path represents a signature.
+   */
+  public boolean isSignature() {
+    return getCoordinates() != null && getCoordinates().getSignatureType() != null;
+  }
+
+  /**
+   * Returns the "main", non-subordinate path of this path. The "main" path is never a hash.
+   */
+  @Nonnull
+  public MavenPath main() {
+    MavenPath mavenPath = this;
+    while (mavenPath.isSubordinate()) {
+      mavenPath = mavenPath.subordinateOf();
+    }
+    return mavenPath;
+  }
+
+  /**
+   * Returns the "parent" path, that this path is subordinate of, or {@code null}.
+   */
+  @Nullable
+  public MavenPath subordinateOf() {
+    Coordinates mainCoordinates = null;
+    if (isHash()) {
+      int hashSuffixLen = hashType.getExt().length() + 1; // the dot
+      if (getCoordinates() != null) {
+        mainCoordinates = new Coordinates(
+            getCoordinates().isSnapshot(),
+            getCoordinates().getGroupId(),
+            getCoordinates().getArtifactId(),
+            getCoordinates().getVersion(),
+            getCoordinates().getBaseVersion(),
+            getCoordinates().getClassifier(),
+            getCoordinates().getExtension().substring(0, getCoordinates().getExtension().length() - hashSuffixLen),
+            getCoordinates().getSignatureType()
```

Comment:
um, why not assign a local to getCoordinates() instead of invoking this method all the time?  Compiler _could_ optimize this, but really given how many times you need to call getCoordinates() it seems to suggest a local to resolve an object that would be better.